### ISSUE TITLE
GPGRecipients_GPGRecipients_test.gpg after introducing check for b:GPGCorrespondingTo

### ISSUE THREAD

**Issue by ThomasAH** [jamessan/vim-gnupg#75]
Changeset e7be11287357641a97cb3b4cb29a66ce33c395a3 (GPGEdit{Recipients,Options}: Check for presence of b:GPGCorrespondingTo) introduces a bug: When opening test.gpg and using :GPGEditRecipients, vim opens one new (empty!) buffer named "GPGRecipients_test.gpg" and a second new buffer named "GPGRecipients_GPGRecipients_test.gpg" which contains the recipient list.

The text in GPGRecipients_GPGRecipients_test.gpg is:

```
GPG: Please edit the list of recipients, one recipient per line.
GPG: Unknown recipients have a prepended "!".
GPG: Lines beginning with "GPG:" are removed automatically.
GPG: Data after recipients between and including "(" and ")" is ignored.
GPG: Closing this buffer commits changes.
GPG: ----------------------------------------------------------------------
GPG: ----------------------------------------------------------------------
GPG: Please edit the list of recipients, one recipient per line.
GPG: Unknown recipients have a prepended "!".
GPG: Lines beginning with "GPG:" are removed automatically.
GPG: Data after recipients between and including "(" and ")" is ignored.
GPG: Closing this buffer commits changes.
GPG: ----------------------------------------------------------------------
Test <test@example.com>             (ID: 0x1122334455667788 created at Wed 07 Dec 2016 4:00:00 PM CET)
```

This happens on Debian jessie (vim 2:7.4.488-7+deb8u1) and Debian wheezy (vim 2:7.3.547-7+deb7u1), even with no ~/.vimrc, and ~/.vim/ only containing plugin/gnupg.vim

**Comment by ThomasAH** [comment]
I have a solution for a different problem and for some reason this fixed this
bug, too:
When using things like :GPGEditRecipients or :GPGViewRecipients on a file that
is not encrypted, b:GPGEncrypted is unset, so the old check did not trigger.

Here is the patch. Sorry, no pull request since I did not use git:

```
# HG changeset patch
# User Thomas Arendsen Hein <thomas@intevation.de>
# Date 1481275774 -3600
#      Fri Dec 09 10:29:34 2016 +0100
# Node ID dd2bf3e845d4c99502ecf0b365fa825fa9a14782
# Parent  79695d985e74f867fa631d94116fb3fd12225961
fix guard for unencrypted files

b:GPGEncrypted was unset (not 0) for unencrypted files.

For an unknown reason this also fixes
https://github.com/jamessan/vim-gnupg/issues/75
(GPGRecipients_GPGRecipients_test.gpg after introducing check for
b:GPGCorrespondingTo)

diff -r 79695d985e74 -r dd2bf3e845d4 plugin/gnupg.vim
--- a/plugin/gnupg.vim	Sun Dec 04 14:07:36 2016 -0500
+++ b/plugin/gnupg.vim	Fri Dec 09 10:29:34 2016 +0100
@@ -645,7 +645,7 @@
   call s:GPGDebug(2, 'called '. auType .'Pre autocommand for ' . autocmd_filename)
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGError
     let blackhole = input("Message could not be encrypted! (Press ENTER)")
     echohl None
@@ -747,7 +747,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGViewRecipients()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -790,7 +790,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGEditRecipients()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -913,7 +913,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGFinishRecipientsBuffer()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -990,7 +990,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGViewOptions()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -1017,7 +1017,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGEditOptions()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -1108,7 +1108,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGFinishOptionsBuffer()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None

```

**Comment by ThomasAH** [comment]
Please do not apply this patch, while it fixes the things I described, it causes the list of recipients not to be updated when closing the edit window. I assume it is needed for some of the lines and the guard should not be used at all (so it was luck it did not work) for other parts.

**Comment by ThomasAH** [comment]
Here is the corrected version, the only difference is checking for
b:GPGCorrespondingTo instead of b:GPGEncrypted in
s:GPGFinishRecipientsBuffer() and s:GPGFinishOptionsBuffer():


```
# HG changeset patch
# User Thomas Arendsen Hein <thomas@intevation.de>
# Date 1481728281 -3600
# Parent  ff380244bbeda996750d2cb40576a493717bd181
fix guard for unencrypted files

1. b:GPGEncrypted was unset (not 0) for unencrypted files.
2. Recipients and options buffer only has b:GPGCorrespondingTo,
   not b:GPGEncrypted

For an unknown reason this also fixes
https://github.com/jamessan/vim-gnupg/issues/75
(GPGRecipients_GPGRecipients_test.gpg after introducing check for
b:GPGCorrespondingTo)

diff -r ff380244bbed plugin/gnupg.vim
--- a/plugin/gnupg.vim	Wed Dec 14 15:55:02 2016 +0100
+++ b/plugin/gnupg.vim	Wed Dec 14 16:13:23 2016 +0100
@@ -645,7 +645,7 @@
   call s:GPGDebug(2, 'called '. auType .'Pre autocommand for ' . autocmd_filename)
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGError
     let blackhole = input("Message could not be encrypted! (Press ENTER)")
     echohl None
@@ -747,7 +747,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGViewRecipients()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -790,7 +790,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGEditRecipients()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -913,7 +913,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGFinishRecipientsBuffer()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGCorrespondingTo"))
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -990,7 +990,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGViewOptions()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -1017,7 +1017,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGEditOptions()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGEncrypted") || b:GPGEncrypted == 0)
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
@@ -1108,7 +1108,7 @@
   call s:GPGDebug(3, ">>>>>>>> Entering s:GPGFinishOptionsBuffer()")
 
   " guard for unencrypted files
-  if (exists("b:GPGEncrypted") && b:GPGEncrypted == 0)
+  if (!exists("b:GPGCorrespondingTo"))
     echohl GPGWarning
     echom "File is not encrypted, all GPG functions disabled!"
     echohl None
```

**Comment by jamessan** [comment]
Thanks for continuing to look at this!  I do see some possible improvements to make to the `Finish` functions.  However, I still can't reproduce your original problem.  Are you able to provide a clear set of steps starting from `vim -u NONE -N -c 'source plugin/gnupg.vim'` that can reproduce the issue?

I'm also not sure that the proposed changes are appropriate.  Requiring the `b:GPGEncrypted` variable exists is intentional.  Treating the lack of it as a problem is a big behavioral change.

**Comment by ThomasAH** [comment]
Yes, I can reproduce it with a clean vim, first terminal contains the commands I used, second terminal is test.gpg (new file, but same thing happens with existing files), third is an unencrypted file (again new, but does not matter either):
![vim-gnupg-issue75](https://cloud.githubusercontent.com/assets/4928697/21216102/2fdae4f2-c2a5-11e6-9c3e-3362cf63712f.png)

**Comment by ThomasAH** [comment]
> I'm also not sure that the proposed changes are appropriate. Requiring the b:GPGEncrypted variable exists is intentional. Treating the lack of it as a problem is a big behavioral change.

The variable is only guaranteed to exist for encrypted buffers. Your code currently does:
  `if b:GPGEncrypted exists, and is zero: bail out with an "file is not encrypted" error`
but it should do:
  `if b:GPGEncrypted does not exist, or (implicit "exists and is") zero: bail out with an "file is not encrypted" error`

b:GPGEncrypted does not exist for unencrypted buffers, so in this case a value of 0 should be assumed.

For the GPGFinish... functions, checking for b:GPGEncrypted is not helpful, because they are not encrypted (despite having a .gpg extension), but they have b:GPGCorrespondingTo. I was thinking about checking `getbufvar(b:GPGCorrespondingTo, "b:GPGEncrypted")`, but this would be overkill.

**Comment by jamessan** [comment]
`vim -u NONE -N -c 'source plugin/gnupg.vim | GPGEditRecipients' foo.gpg`

This isn't right.  It's the wrong order of operation.  `-c` runs after the first file has been loaded.  What's happening is that vim opens foo.gpg as a normal file, then you load gnupg.vim.  Now you run `GPGEditRecipients` which ends up triggering the autocommands for a gpg file on the GPGEditRecipients buffer.

This shouldn't be happening at all in normal use, so I have to assume you're using some sort of plugin manager and telling it to lazily load gnupg.vim, which is causing these problems.

**Comment by ThomasAH** [comment]
You told me to use -c source :)

And the behavior was identical to renaming ~/.vimrc and ~/.vim/, creating a new ~/.vim/plugin/ directory, copying gnupg.vim to this new directory, and finally opening test.gpg

I am not actively using a plugin manager (though the Debian package "vim-addon-manager" is installed on this system), but shouldn't "-u NONE" completely disable it?

And I additionally tried:
`vim -u NONE -N -c 'source vim-gnupg/plugin/gnupg.vim'`
followed by ":edit test.gpg" (which I created before):
This again opens three vim windows: test.gpg, GPGRecipients_test.gpg and GPGRecipients_GPGRecipients_test.gpg

And now I tried it on a completely different system:
Debian wheezy, 32bit (not amd64 like the one above), no NFS-mounted $HOME (unlike the one I used above), the only vim-related packages are:
```
$ dpkg -l '*vim*'|grep ^i
ii  vim            2:7.3.547-7+ i386         Vi IMproved - enhanced vi editor
ii  vim-common     2:7.3.547-7+ i386         Vi IMproved - Common files
ii  vim-runtime    2:7.3.547-7+ all          Vi IMproved - Runtime files

```
(this is 2:7.3.547-7+deb7u1)

Still the same behavior.

These latest tests were done with current git/master:
```
commit dda14abb3c51166553015ef489e9ac9acb859904
Author: James McCoy <jamessan@jamessan.com>
Date:   Wed Dec 14 23:05:40 2016 -0500

$ sha1sum gnupg.vim
721a996c4304669fc3d151a95f4a6122e9198e64  gnupg.vim
```

The tests in earlier posts were done with the current git/master of that time:
```
commit e7be11287357641a97cb3b4cb29a66ce33c395a3
Author: James McCoy <jamessan@jamessan.com>
Date:   Sun Dec 4 13:56:39 2016 -0500
```

Can you verify that we are testing the same version and settings? (no ~/.vimrc, identical gnupg.vim)

**Comment by ThomasAH** [comment]
And even if you can't reproduce it yourself, can you comment on my reasoning in https://github.com/jamessan/vim-gnupg/issues/75#issuecomment-267265616 about why I think my patch does the correct thing?

**Comment by jamessan** [comment]
> can you comment on my reasoning in #75 (comment) about why I think my patch does the correct thing?

Your assumptions in that comment are incorrect.  `b:GPGEncrypted` _only_ exists if there is a file corresponding to the buffer name.  The intention is that `vim existing_unencrypted.asc` continues to treat that file as an unencrypted file.  However, `vim nonexistent.asc` will encrypt the contents of the buffer when it writes out the new file `nonexistent.asc`.

Also, `s:GPGEditRecipents` and `s:GPGEditOptions` start off in the encrypted buffer, so the initial checks for `b:GPGEncrypted` are looking at the correct buffer.

The problem is your scenario is that either

* `gnupg.vim` is loaded after the foo.gpg buffer has been created
* The command that creates the foo.gpg buffer is doing so with auto-commands disabled (e.g., `noau edit foo.gpg`), so `s:GPGInit` is never called.

**Comment by jamessan** [comment]
I think the original issue here is likely also fixed by e2aec69d2d0f49dbc61916eadd17c8868f077fe4, so I'm going to close it.  Please let me know if you have any more issues.
